docs/specs/ldap.md

```diff
@@ -0,0 +1,133 @@
+
+When the connection URL starts with `ldap://`, StartTLS will be tried.
+If the LDAP server responds protocolError, as defined in [Section 4.14.1 in RFC4511](https://datatracker.ietf.org/doc/html/rfc4511#section-4.14.1),
+then Authgear will treat it as no TLS is required.
+
+When the connection URL starts with `ldaps://`, then Authgear will connect to the LDAP server with TLS directly, without using StartTLS.
+
+## Authenticate Authgear as a LDAP client
+
+It is very common that before a LDAP client can run any [Search Operation](https://datatracker.ietf.org/doc/html/rfc4511#section-4.5),
+the LDAP client must perform [Bind Operation](https://datatracker.ietf.org/doc/html/rfc4511#section-4.2) first.
+
+Authgear supports Simple Bind with username and password.
+
+## Configuration of LDAP servers
+
+The following example demonstrates how the developer should configure Authgear to connect to their LDAP servers.
+
+In `authgear.yaml`
+
+```yaml
+identity:
+  ldap:
+    servers:
+    - name: ldap1
+      url: "ldap://localhost:389"
+      base_distinguished_name: "dc=localhost"
+      relative_distinguished_name_attribute: "uid"
+    - name: ldap2
+      url: "ldap://mycompany.com:389"
+      base_distinguished_name: "dc=mycompany,dc=com"
+      relative_distinguished_name_attribute: "uid"
+```
+
+- `identity.ldap.servers.name`: A name that only exists in `authgear.yaml` and `authgear.secrets.yaml` for associating a LDAP server. It serves no other purpose.
+- `identity.ldap.servers.url`: The connection URL to the LDAP server. The scheme MUST be `ldap:` or `ldaps:`. The URL MUST contain `host`, and optionally a port. If the port is omitted, the default port of the scheme is assumed. The default port of `ldap:` is `389`, while the default port of `ldaps:` is `636`. The URL MUST NOT contain other elements, such as path, nor query.
+
+> Why does `identity.ldap.servers.url` allow scheme, host, and port?
+> The LDAP URL, defined in [Section 2 in RFC 4516](https://datatracker.ietf.org/doc/html/rfc4516#section-2), is syntactically different from the URL defined in [RFC3986](https://datatracker.ietf.org/doc/html/rfc3986).
+> In particular, a LDAP URL can contain multiple question mark characters.
+> To ease implementation, we do not support the LDAP URL, and require a RFC3986 URL (which is implemented by the standard library net/url package).
+
+- `identity.ldap.servers.base_distinguished_name`: The base distinguished name to construct a Search Request, as defined in [Section 4.5.1 in RFC4511](https://datatracker.ietf.org/doc/html/rfc4511#section-4.5.1).
+- `identity.ldap.servers.relative_distinguished_name_attribute`: The attribute name Authgear should use to construct the search request. For example, if the value is `uid`, and the end-user gives a username of `user1`, and the base distinguished name is `dc=example,dc=com`, then the relative distinguished name is `uid=user1`, and the distinguished name is `uid=user1,dc=example,dc=com`.
```

Comment:
I think we need to do a search request to get user's DN in ldap rather just prepend `uid=user1` to the the base DN.

From what I read, usually there will be multiple entry nodes between the root and the entry representing the user in the directory information tree. i.e. Usually there will be one or more organization units in between so that the DN usually would look like `cn=Some User,ou=Department Name,ou=A Region,dc=example,dc=com`

I am not so sure we need to document this or not. Or is it already the implementation detail? 🤔

Reply:
In your example `cn=Some User,ou=Department Name,ou=A Region,dc=example,dc=com`, the `base_distinguished_name` is `ou=Department Name,ou=A Region,dc=example,dc=com`. This base DN is configured by the developer.

What we collect from the end-user are the username and the password. We need to combine the username with the base DN, so I design `relative_distinguished_name_attribute`.

Reply:
What if users are actually from two different departments and different region? 
.e.g.
1. `cn=A Developer,ou=Engineering Team,ou=Hong Kong,dc=example,dc=com`
2. `cn=A Design,ou=Design Team,ou=United Kingdom,dc=example,dc=com`

In such case, we cannot just simple combine end user input and the base DN. Also, end user input (i.e. `uid` or `email`) may not appears in the DN in the directory. It is up to ldap admin to configure to use either `cn` or other rdn like `uid` to representing the DN